art/attacks/evasion/pixel_threshold.py

```diff
@@ -0,0 +1,1346 @@
+        :type  es:            `int`
+        :param targeted:      Indicates whether the attack is targeted (True)
+                              or untargeted (False)
+        :type  targeted:      `bool`
+        :param verbose:       Indicates whether to print verbose messages of ES used
+        :type  verbose:       `bool`
+        """
+        super(PixelThreshold, self).__init__(classifier)
+        self._project = True
+        kwargs = {'th': th, 'es': es, 'targeted': targeted, 'verbose': verbose}
+        PixelThreshold.set_params(self, **kwargs)
+        self.type_attack = -1
+
+    def set_params(self, **kwargs):
+        """
+        Take in a dictionary of parameters and applies attack-specific checks
+        before saving them as attributes.
+        """
+        super(PixelThreshold, self).set_params(**kwargs)
+
+        if self.th is not None:
+            if self.th <= 0:
+                raise ValueError(
+                    'The perturbation size `eps` has to be positive.')
+        if not isinstance(self.es, int):
+            raise ValueError('The flag `es` has to be of type int.')
+        if not isinstance(self.targeted, bool):
+            raise ValueError('The flag `targeted` has to be of type bool.')
+        if not isinstance(self.verbose, bool):
+            raise ValueError('The flag `verbose` has to be of type bool.')
+
+        return True
+
+    def generate(self, x, y=None, **kwargs):
+        """
+        Generate adversarial samples and return them in an array.
+
+        :param x: An array with the original inputs.
+        :type x: `np.ndarray`
+        :param y: Target values (class labels) one-hot-encoded of shape
+                  (nb_samples, nb_classes) or indices of shape (nb_samples,).
+                  Only provide this parameter if you'd like to use true labels
+                  when crafting adversarial samples. Otherwise, model
+                  predictions are used as labels to avoid the "label leaking"
+                  effect
+                  (explained in this paper: https://arxiv.org/abs/1611.01236).
+                  Default is `None`.
+        :type y: `np.ndarray`
+        :return: An array holding the adversarial examples.
+        :rtype: `np.ndarray`
+        """
+        y = check_and_transform_label_format(
+            y, self.classifier.nb_classes(), return_one_hot=False)
+
+        if y is None:
+            if self.targeted:
+                raise ValueError(
+                    'Target labels `y` need to be provided for a targeted '
+                    'attack.'
+                )
+            y = np.argmax(self.classifier.predict(x), axis=1)
+        else:
+            if len(y.shape) > 1:
+                y = np.argmax(y, axis=1)
+
+        if self.th is None:
+            logger.info('Performing minimal perturbation Attack.')
+
+        if isinstance(self.classifier, PyTorchClassifier):
+            x = np.swapaxes(x, 1, 3).astype(np.float32)
```

Comment:
Please remove these lines, because `x` is provided by the user in the correct format.

Reply:
No, actually, when PyTorch is used the shape is `(img_channel, img_rows, img_cols) or channels_first`, however, my code requires the image in `channels_last` format. Therefore, I check if the PyTorch is used.

Reply:
Please note that the `get_classifier_pt` expects an input of shape `(1,28,28)` and not `(28,28,1)`
The suggestion would, therefore, create a contradiction to the expected shape.

https://github.com/IBM/adversarial-robustness-toolbox/blob/e6d0b53a16bca37712a78ad916018c67a5951abe/tests/utils_test.py#L570

Reply:
Please note that the `PyTorchClassifier` expects an input of shape `(1,28,28) or Channels First` and not `(28,28,1) or Channels Last`. Therefore, input would be in the shape `(1,28,28) or Channels First`. However, my attacks use the format `(28,28,1) or Channels Last` and these lines change it to the required format.

Reply:
Ok, in that case please check for the value of `classifier.channel_index` to be 1 or 3 instead of checking for `isinstance(self.classifier, PyTorchClassifier)` because your attack could also be used, for example, with `KerasClassifier` and channel_index=1.

Reply:
I see, indeed checking for the `classifier.channel_index` makes more sense then checking for `class: PyTorchClassifier`. 
I have updated the checking.

The answer for the question is mentioned in the comment.